### PR TITLE
ref(replays): Refactor the ReplayTable to make it more reusable

### LINKED ISSUE #40726: Design Bug: On narrow viewports the Replay Index 'session' column becomes un-clickable
We should be dropping columns to keep the 'session' column visible.

You literally can't click into replays because the link inside the 1st column is truncated away.

Image

### PR DESCRIPTION
**Pull request by ryan953**
Refactor the ReplayTable so it's more re-usable for the upcoming `Other Session` tab that's going into the replay details page.

Previously we had this pattern of setting `showFieldXYZ` props whenever we wanted to see a column or not. But only some columns were guarded by that, they were also guarded by the total page width. This combo made it hard to reason about what column was going to be shown, and if list were put into a container that's smaller than the window we would not get those columns hidden. 

So I created the new `VisibleColumns` type and props so each callsite can be explicit about which columns it wants to see, then i hoisted up the media query check into all the existing callsites, so they can continue to work the way they have before. 

We're left with the new ReplayTable, which is mostly concerned with sorting out which columns to show, then delegates to HeaderCell and TableCell to get those columns rendered. 

Fixes #40726

### CODE REVIEW

File: static/app/views/organizationGroupDetails/groupReplays/groupReplays.tsx [hunk 1]
```diff
@@ -129,11 +141,17 @@ const GroupReplaysTable = ({
...
-        showProjectColumn={false}
         sort={first(eventView.sorts)}
-        fetchError={fetchError}
+        visibleColumns={{
+          activity: true,
+          countErrors: true,
+          duration: true,
+          session: true,
+          startedAt: hasRoomForColumns,
+        }}
```
Comment (billyvg) [inline static/app/views/organizationGroupDetails/groupReplays/groupReplays.tsx:154]: 👍

File: static/app/views/replays/replayTable/index.tsx [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    visibleColumns.session ? <HeaderCell column="session" sort={sort} /> : null,
+    visibleColumns.projectId ? <HeaderCell column="projectId" sort={sort} /> : null,
+    visibleColumns.slowestTransaction ? (
+      <HeaderCell column="slowestTransaction" sort={sort} />
+    ) : null,
+    visibleColumns.startedAt ? <HeaderCell column="startedAt" sort={sort} /> : null,
+    visibleColumns.duration ? <HeaderCell column="duration" sort={sort} /> : null,
+    visibleColumns.countErrors ? <HeaderCell column="countErrors" sort={sort} /> : null,
+    visibleColumns.activity ? <HeaderCell column="activity" sort={sort} /> : null,
+  ].filter(Boolean);
```
Comment (billyvg) [inline static/app/views/replays/replayTable/index.tsx:44]: This assumes HeaderCell.column == keyof visibleColumns...

```suggestion
  const tableHeaders = Object.entries(visibleColumns).filter(([, visible]) => visible).map(([column]) => <HeaderCell column={column} sort={sort});
```

File: static/app/views/replays/replayTable/tableCell.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  referrer: string;
+  replay: ReplayListRecord | ReplayListRecordWithTx;
+};
+
+function TableCell({column, eventView, organization, referrer, replay}: Props) {
+  const location = useLocation();
+  const {projects} = useProjects();
+  const project = projects.find(p => p.id === replay.projectId);
+
+  switch (column) {
```
Comment (billyvg) [inline static/app/views/replays/replayTable/tableCell.tsx:35]: Feels like each of these column types should be moved to a subcomponent

**Review by billyvg — APPROVED** [review]
Nothing blocking

File: static/app/views/replays/replayTable/index.tsx (hunk 2 above)
Comment (ryan953) [reply static/app/views/replays/replayTable/index.tsx:44]: I'm worried about the display order of this.... object keys are not technically sorted. 

maybe I can switch the visibleColumns object into an array though.

File: static/app/views/replays/replayTable/tableCell.tsx (hunk 3 above)
Comment (ryan953) [reply static/app/views/replays/replayTable/tableCell.tsx:35]: yeah... the list inside of replayTable.tsx is kind of calling for that too with all the `column="foo"` props

File: static/app/views/replays/replayTable/index.tsx (hunk 2 above)
Comment (ryan953) [reply static/app/views/replays/replayTable/index.tsx:52]: switched!

File: static/app/views/replays/replayTable/tableCell.tsx (hunk 3 above)
Comment (ryan953) [reply static/app/views/replays/replayTable/tableCell.tsx]: changed!

File: static/app/views/replays/replayTable/index.tsx (hunk 2 above)
Comment (billyvg) [reply static/app/views/replays/replayTable/index.tsx]: I was curious if this was still the case re keys in insertion order and it looks like in es2015, only certain methods (`keys()` being excluded) would guarantee order for non-int keys. But as of es2020, all of them should guarantee order.

**Review by billyvg — APPROVED** [review]

**Pull request merged**